src/coreclr/src/vm/processdiagnosticsprotocolhelper.cpp

```diff
@@ -0,0 +1,214 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "common.h"
+#include "fastserializer.h"
+#include "processdiagnosticsprotocolhelper.h"
+#include "eventpipeeventsource.h"
+#include "diagnosticsprotocol.h"
+#include "diagnosticserver.h"
+
+#ifdef FEATURE_PERFTRACING
+
+static inline uint32_t GetStringLength(const WCHAR *&value)
+{
+    return static_cast<uint32_t>(wcslen(value) + 1);
+}
+
+static bool TryWriteString(uint8_t * &bufferCursor, uint16_t &bufferLen, const WCHAR *&value)
+{
+    uint32_t stringLen = GetStringLength(value);
+    S_UINT16 totalStringSizeInBytes(stringLen * sizeof(WCHAR) + sizeof(uint32_t));
+    ASSERT(!totalStringSizeInBytes.IsOverflow());
+    ASSERT(bufferLen >= totalStringSizeInBytes.Value());
+    if (bufferLen < totalStringSizeInBytes.Value() || totalStringSizeInBytes.IsOverflow())
+        return false;
+
+    memcpy(bufferCursor, &stringLen, sizeof(stringLen));
+    bufferCursor += sizeof(stringLen);
+
+    memcpy(bufferCursor, value, stringLen * sizeof(WCHAR));
+    bufferCursor += stringLen * sizeof(WCHAR);
+
+    bufferLen -= totalStringSizeInBytes.Value();
+    return true;
+}
+
+uint16_t ProcessInfoPayload::GetSize()
+{
+    LIMITED_METHOD_CONTRACT;
+
+    // The protocol buffer is defined as:
```

Comment:
We need an update to the spec, ideally before or together with the implementation : )
Once the spec is updated these large comments could be replaced with a link to the relevant section of the spec on github.